File: src/boil/detectors/c_unknown_type.go
// Detector for C unknown type name errors.
package detectors

// CUnknownTypeDetector detects C unknown type name errors
type CUnknownTypeDetector struct {
	*BaseDetector
}

func NewCUnknownTypeDetector() (*CUnknownTypeDetector, error) {
	base, err := NewBaseDetector(
		"CUnknownTypeDetector",
		100,
		map[string]string{
			"unknown_type_name": `(?P<file_path>[a-zA-Z0-9_./\-]+\.c):(?P<line_number>\d+):\d+:\s+error:\s+unknown type name\s+[''](?P<type_name>[^'']+)['']`,
		},
		[]DetectorExample{
			{
				Name:     "unknown_type_size_t",
				Input:    "utils.c:8:1: error: unknown type name 'size_t'",
				ClueType: "unknown_type_name",
				Context: map[string]string{
					"file_path":   "utils.c",
					"line_number": "8",
					"type_name":   "size_t",
				},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &CUnknownTypeDetector{base}, nil
}

File: src/pipeline/detectors/c_unknown_type.py
"""
Detector for C "unknown type name" errors.
"""

import re
import typing as T
from src.pipeline.detectors.base import Detector
from src.pipeline.models import ErrorClue


class CUnknownTypeDetector(Detector):
    """
    Detect C "unknown type name" errors.

    Matches patterns like:
    - error: unknown type name 'TSNode'
    - error: unknown type name 'struct winsize'

    The planner will use git grep to find which header file defines the type.
    """

    PATTERNS = {
        "unknown_type_name": r"(?P<file_path>[a-zA-Z0-9_./\-]+\.c):(?P<line_number>\d+):\d+:\s+error:\s+unknown type name\s+['\u2018](?P<type_name>[^'\u2019]+)['\u2019]",
    }

    EXAMPLES = [
        (
            "dim.c:17:1: error: unknown type name 'TSLanguage'",
            {
                "clue_type": "unknown_type_name",
                "confidence": 1.0,
                "context": {
                    "file_path": "dim.c",
                    "line_number": "17",
                    "type_name": "TSLanguage",
                },
            },
        ),
        (
            "test.c:10:3: error: unknown type name 'struct winsize'",
            {
                "clue_type": "unknown_type_name",
                "confidence": 1.0,
                "context": {
                    "file_path": "test.c",
                    "line_number": "10",
                    "type_name": "struct winsize",
                },
            },
        ),
    ]
